### ISSUE TITLE
NUnit 3.x support

### ISSUE THREAD

**Issue by damageboy** [fsprojects/FsUnit#82]
Pretty please? Is there anything to do here beyond bumping the version?

**Comment by sergey-tihon** [comment]
Yes, we need `FAKE` support of `NUnit2` first
https://github.com/fsharp/FAKE/issues/1010

**Comment by sergey-tihon** [comment]
So, probably it is time to try support of NUnit 3 https://github.com/fsharp/FAKE/pull/1064

**Comment by sergey-tihon** [comment]
@et1975 do you have experience of writing custom constrains for `NUnit3`?

**Comment by et1975** [comment]
I wanted to take a look at v3 anyways, might as well port this.

**Comment by sergey-tihon** [comment]
Thank you @et1975

**Comment by et1975** [comment]
Once nunit/nunit#1196 is resolved the changes in et1975/FsUnit/equals branch should do it.

**Comment by rprouse** [comment]
The NUnit team is working on custom message formatters based on the issue nunit/nunit#1196 and original PR by @et1975. @CharliePoole has create PR nunit/nunit#1222

This fix is intended for you, so we would appreciate a code review to see if it meets your needs.

Once it is merged, if you need to test using the NuGet packages produced by our continuous build, I can point you to them.

@sergey-tihon what do you need custom constraints for? They are fairly easy to write. If you need help, please reach out.

**Comment by CharliePoole** [comment]
Here's another question about our (NUnit's) support for F#.

Can anybody think of a way to add a feature to the NUnit framework, so that it is **only** accessible to F#. Resolving nunit/nunit#1196 with a static global is cool for F# users, who will access it in a top-level module, rather than from a test. It's a bit more dangerous in C# (not to mention VB) where calling it from within a test is the most natural way to do it. Adding a new formatter inside a test using the a static method creates a nasty side-effect.

Maybe we need a separate module per language on top of the framework, but that's a very busy solution. Which is what leads to the question: how to give something to F# users without giving it to other users?

**Comment by sergey-tihon** [comment]
@rprouse 
Would be nice to test in once it is merged. Appreciate if you point me to you nightly packages.
There are only two custom constrains in current codebase:
- [EqualsConstraint](https://github.com/fsprojects/FsUnit/blob/master/src/FsUnit.NUnit/FsUnit.fs#L10) - we use it to pretty-print F# values
- [ChoiceConstraint](https://github.com/fsprojects/FsUnit/blob/master/src/FsUnit.NUnit/FsUnit.Extra.fs#L6) which allow us to write constrains for F# [Core.Choice<'T1,'T2>](https://msdn.microsoft.com/en-us/library/ee353439.aspx) values, like
```fsharp
Choice<int, string>.Choice1Of2(42) |> should be (choice 1)
```

@CharliePoole I think that the reason why @et1975 did not want to do it in `OneTimeSetUp` is that we can do it once per assembly with static method. 
Users of `FsUnit` can get it for free because it will be in `FsUnit.dll` and that is enough.

**Comment by CharliePoole** [comment]
@sergey-tihon It sounds like F# users are likely to be satisfied by this. My concern is that we are exposing the feature to all users and those working in other languages who use this in the most "obvious" way will be unexpectedly impacting global state. That is, such users are likely to call the method in a setup or onetimesetup. They will expect the custom format to go away when the scope of the setup terminates but that won't happen.

In fact F# users could fall into the same trap but probably won't because they have an easier way to use the feature.

I'll keep playing with this and let you know when there is something relatively final to test.

**Comment by rprouse** [comment]
@sergey-tihon the NuGet feed is at https://ci.appveyor.com/nuget/nunit

Currently you need to figure out which build a fix is in by looking at the AppVeyor builds and getting the build number. Less than ideal. We plan to fix it, but haven't yet. Remind me once the PR is merged and I will get you a package version that will work for you.

I assume that the custom formatter will solve your need for the EqualsConstraint and open up more of our constraints for your use. Looking at ChoiceConstraint, it looks like it should just work, although my F# foo is not strong :smile:

**Comment by rprouse** [comment]
@sergey-tihon if you wanted to test with the NuGet package of the current PR, the latest version is `3.1.0-CI-2072`

**Comment by CharliePoole** [comment]
@sergey-tihon Note however that the current PR is marked Do Not Merge because it is still in development. The calling sequence could change.

**Comment by sergey-tihon** [comment]
Support of NUnit 3 released in [FsUnit 2.0.0](https://www.nuget.org/packages/FsUnit/2.0.0)
Let's tack support of NUnit v3.2 in separate issue - https://github.com/fsprojects/FsUnit/issues/86
